File: backend/src/test/java/com/poeflipfinder/backend/usecase/computeflipopportunities/BulkBuyOpportunityFinderTest.java
package com.poeflipfinder.backend.usecase.computeflipopportunities;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.within;

import com.poeflipfinder.backend.entity.Currency;
import com.poeflipfinder.backend.entity.ExchangeMarketSnapshot;
import com.poeflipfinder.backend.entity.FlipOpportunity;
import com.poeflipfinder.backend.entity.League;
import com.poeflipfinder.backend.entity.Technique;
import java.time.Instant;
import java.util.List;
import org.junit.jupiter.api.Test;

class BulkBuyOpportunityFinderTest {

    private static final Instant NOW = Instant.parse("2026-08-06T12:00:00Z");

    private final League league = new League(1L, "Standard", "Standard", false, true);
    private final Currency chaos =
            new Currency(1L, "Metadata/Items/Currency/CurrencyRerollRare", "Chaos Orb", null, Currency.ItemType.CURRENCY);
    private final Currency divine =
            new Currency(2L, "Metadata/Items/Currency/CurrencyModValues", "Divine Orb", null, Currency.ItemType.CURRENCY);
    private final Currency deck = new Currency(
            3L, "Metadata/Items/DivinationCards/DivinationCardStackedDeck", "Stacked Deck", null,
            Currency.ItemType.DIVINATION_CARD);
    private final Currency wisdom = new Currency(
            4L, "Metadata/Items/Currency/CurrencyIdentification", "Scroll of Wisdom", null, Currency.ItemType.CURRENCY);

    private final BulkBuyOpportunityFinder finder = new BulkBuyOpportunityFinder();
    private final DivineChaosRate rate210 = new DivineChaosRate(chaos, divine, 210.0);

    private ExchangeMarketSnapshot snapshot(
            Currency currencyA,
            Currency currencyB,
            double lowestRatioA,
            double lowestRatioB,
            double highestRatioA,
            double highestRatioB,
            long lowestStockA,
            long highestStockA,
            long lowestStockB,
            long highestStockB) {
        return new ExchangeMarketSnapshot(
                1L, 999L, league, currencyA, currencyB, NOW, 100, 100,
                lowestStockA, highestStockA, lowestStockB, highestStockB,
                lowestRatioA, highestRatioA, lowestRatioB, highestRatioB);
    }

    @Test
    void find_bothDirectionsViable_producesTwoOpportunitiesWithHandVerifiedNumbers() {
        // Chaos-deck leg: priceAtLowest=8, priceAtHighest=13 -> buy=floor(13)-1=12, sell=floor(8)+1=9.
        ExchangeMarketSnapshot chaosLeg = snapshot(chaos, deck, 1, 8, 1, 13, 100, 50, 1, 1);
        // Divine-deck leg: priceAtLowest=1700, priceAtHighest=1900 -> buy=1899, sell=1701.
        ExchangeMarketSnapshot divineLeg = snapshot(divine, deck, 1, 1700, 1, 1900, 30, 80, 1, 1);

        List<FlipOpportunity> opportunities = finder.find(List.of(chaosLeg, divineLeg), rate210);

        assertThat(opportunities).hasSize(2);
        FlipOpportunity chaosToDivine =
                opportunities.stream().filter(o -> o.sell().get(0).currency().equals(divine)).findFirst().orElseThrow();
        FlipOpportunity divineToChaos =
                opportunities.stream().filter(o -> o.sell().get(0).currency().equals(chaos)).findFirst().orElseThrow();

        // Rescaled so the trade always ends at exactly 1 Divine sold instead
        // of starting from 1 Chaos ("1 Chaos -> 0.005 Divine" is a
        // meaningless fraction to read): via always equals the divine leg's
        // sell price (1701, since that's by definition how much of the
        // intermediary sells for exactly 1 Divine), and start is however
        // much Chaos that took (1701/12 = 141.75). Margin is unaffected
        // (scale-invariant); profit scales up proportionally (0.481481 * 141.75 = 68.25).
        assertThat(chaosToDivine.technique()).isEqualTo(Technique.BULK_BUY);
        assertThat(chaosToDivine.start().get(0).currency()).isEqualTo(chaos);
        assertThat(chaosToDivine.start().get(0).quantity()).isCloseTo(141.75, within(0.001));
        assertThat(chaosToDivine.via().get(0).currency()).isEqualTo(deck);
        assertThat(chaosToDivine.via().get(0).quantity()).isEqualTo(1701.0);
        assertThat(chaosToDivine.sell().get(0).quantity()).isEqualTo(1.0);
        assertThat(chaosToDivine.marginPercent()).isCloseTo(13.0 / 27 * 100, within(0.001));
        assertThat(chaosToDivine.profit().currency()).isEqualTo(chaos);
        assertThat(chaosToDivine.profit().quantity()).isCloseTo(68.25, within(0.001));
        assertThat(chaosToDivine.volume()).isEqualTo(30); // min(chaos buyLegStock=50, divine sellLegStock=30)

        assertThat(divineToChaos.start().get(0).currency()).isEqualTo(divine);
        assertThat(divineToChaos.start().get(0).quantity()).isEqualTo(1.0);
        assertThat(divineToChaos.via().get(0).quantity()).isEqualTo(1899.0);
        assertThat(divineToChaos.sell().get(0).quantity()).isCloseTo(211.0, within(1e-7));
        assertThat(divineToChaos.marginPercent()).isCloseTo(100.0 / 210, within(0.001));
        assertThat(divineToChaos.profit().currency()).isEqualTo(chaos);
        assertThat(divineToChaos.profit().quantity()).isCloseTo(1.0, within(1e-7));
        assertThat(divineToChaos.volume()).isEqualTo(80); // min(divine buyLegStock=80, chaos sellLegStock=100)
    }

    @Test
    void find_chaosBuyLegNotViable_onlyDivineToChaosDirectionSurvives() {
        // Chaos-wisdom leg: priceAtLowest=1, priceAtHighest=1.5 -> buy floors to 0 (invalid),
        // but sell=floor(1)+1=2 is still valid -- this is exactly the asymmetry
        // UndercutQuote exists to expose: a naive "drop both legs if either
        // quote is unusable" would incorrectly return 0 opportunities here.
        ExchangeMarketSnapshot chaosLeg = snapshot(chaos, wisdom, 2, 2, 2, 3, 40, 70, 1, 1);
        ExchangeMarketSnapshot divineLeg = snapshot(divine, wisdom, 1, 50, 1, 80, 30, 80, 1, 1);

        List<FlipOpportunity> opportunities = finder.find(List.of(chaosLeg, divineLeg), rate210);

        assertThat(opportunities).hasSize(1);
        FlipOpportunity opportunity = opportunities.get(0);
        assertThat(opportunity.start().get(0).currency()).isEqualTo(divine);
        assertThat(opportunity.sell().get(0).currency()).isEqualTo(chaos);
        assertThat(opportunity.via().get(0).quantity()).isEqualTo(79.0); // floor(80)-1
        assertThat(opportunity.sell().get(0).quantity()).isCloseTo(39.5, within(1e-7)); // 79/2
    }

    @Test
    void find_rateUnavailable_returnsEmptyRegardlessOfSnapshots() {
        ExchangeMarketSnapshot chaosLeg = snapshot(chaos, deck, 1, 8, 1, 13, 100, 50, 1, 1);
        ExchangeMarketSnapshot divineLeg = snapshot(divine, deck, 1, 1700, 1, 1900, 30, 80, 1, 1);

        assertThat(finder.find(List.of(chaosLeg, divineLeg), null)).isEmpty();
    }

    @Test
    void find_noMatchingDivineLegForTheSameIntermediary_isNotACandidate() {
        ExchangeMarketSnapshot chaosLeg = snapshot(chaos, deck, 1, 8, 1, 13, 100, 50, 1, 1);

        assertThat(finder.find(List.of(chaosLeg), rate210)).isEmpty();
    }

    @Test
    void find_onlyTheChaosDivineReferencePairItself_excludesItAsACandidate() {
        ExchangeMarketSnapshot chaosDivine = snapshot(chaos, divine, 210, 1, 210, 1, 50, 60, 50, 60);

        assertThat(finder.find(List.of(chaosDivine), rate210)).isEmpty();
    }

    @Test
    void find_nonFiniteRatioOnOneLeg_dropsThatIntermediaryEntirely() {
        ExchangeMarketSnapshot chaosLeg = snapshot(chaos, deck, 0, 8, 1, 13, 100, 50, 1, 1); // lowestRatioA=0
        ExchangeMarketSnapshot divineLeg = snapshot(divine, deck, 1, 1700, 1, 1900, 30, 80, 1, 1);

        assertThat(finder.find(List.of(chaosLeg, divineLeg), rate210)).isEmpty();
    }
}

File: app/test/poe_flip_finder/bulk_buy_opportunity_finder_test.exs
defmodule PoeFlipFinder.BulkBuyOpportunityFinderTest do
  use ExUnit.Case, async: true

  alias PoeFlipFinder.{
    BulkBuyOpportunityFinder,
    Currency,
    DivineChaosRate,
    ExchangeMarketSnapshot,
    League
  }

  # Ported 1:1 from the Java BulkBuyOpportunityFinderTest -- these use cases
  # are already drilled and confirmed traceable to docs/PRD.md § 7.2/7.5
  # (see docs/ELIXIR_TEST_MANIFESTO.md § Use-Case Discovery): buy-competitive
  # / sell-market split, reference trade scaling, zero-volume/non-finite
  # guards, and the base-pair exclusion are all already documented there.

  @now ~U[2026-08-06 12:00:00Z]

  @league %League{
    id: 1,
    external_id: "Standard",
    display_name: "Standard",
    is_current: false,
    has_exchange_activity: true
  }
  @chaos %Currency{
    id: 1,
    external_id: "Metadata/Items/Currency/CurrencyRerollRare",
    display_name: "Chaos Orb",
    category: :currency
  }
  @divine %Currency{
    id: 2,
    external_id: "Metadata/Items/Currency/CurrencyModValues",
    display_name: "Divine Orb",
    category: :currency
  }
  @deck %Currency{
    id: 3,
    external_id: "Metadata/Items/DivinationCards/DivinationCardStackedDeck",
    display_name: "Stacked Deck",
    category: :cards
  }
  @wisdom %Currency{
    id: 4,
    external_id: "Metadata/Items/Currency/CurrencyIdentification",
    display_name: "Scroll of Wisdom",
    category: :currency
  }

  @rate_210 %DivineChaosRate{
    chaos_currency: @chaos,
    divine_currency: @divine,
    chaos_per_divine: 210.0
  }

  defp snapshot(currency_a, currency_b, ratios, stocks) do
    {lowest_ratio_a, lowest_ratio_b, highest_ratio_a, highest_ratio_b} = ratios
    {lowest_stock_a, highest_stock_a, lowest_stock_b, highest_stock_b} = stocks

    %ExchangeMarketSnapshot{
      id: 1,
      generation_id: 999,
      league: @league,
      currency_a: currency_a,
      currency_b: currency_b,
      snapshot_hour: @now,
      volume_traded_a: 100,
      volume_traded_b: 100,
      lowest_stock_a: lowest_stock_a,
      highest_stock_a: highest_stock_a,
      lowest_stock_b: lowest_stock_b,
      highest_stock_b: highest_stock_b,
      lowest_ratio_a: lowest_ratio_a,
      highest_ratio_a: highest_ratio_a,
      lowest_ratio_b: lowest_ratio_b,
      highest_ratio_b: highest_ratio_b
    }
  end

  test "divine-to-chaos opportunity computed with hand-verified numbers" do
    # Chaos-deck leg: price_at_lowest=8, price_at_highest=13 -> buy=floor(13)-1=12,
    # market_sell_price = floor(13) = 13 -- the SAME extreme as the buy price
    # (the worse-for-a-seller one), not the round-trip-favorable 8. This is
    # the exact shape of a real production bug: a wide in-hour spread (here
    # 8 vs 13) previously fed the optimistic 8 into a one-directional sell,
    # overstating how much a real seller could get.
    chaos_leg = snapshot(@chaos, @deck, {1, 8, 1, 13}, {100, 50, 1, 1})

    # Divine-deck leg: price_at_lowest=1700, price_at_highest=1900 -> buy=1899.
    divine_leg = snapshot(@divine, @deck, {1, 1700, 1, 1900}, {30, 80, 1, 1})

    [opportunity] = BulkBuyOpportunityFinder.find([chaos_leg, divine_leg], @rate_210)

    assert opportunity.technique == :bulk_buy
    # Sells through the chaos leg's market price (13, not the
    # round-trip-favorable 8), so 1899 deck / 13 = 146.077 Chaos -- below
    # the 210 c/div direct baseline, i.e. this specific pair is a real loss
    # here, not a profitable flip -- the math still has to be reported
    # accurately either way.
    assert hd(opportunity.start).currency == @divine
    assert hd(opportunity.start).quantity == 1.0
    assert hd(opportunity.via).currency == @deck
    assert hd(opportunity.via).quantity == 1899.0
    assert hd(opportunity.sell).currency == @chaos
    assert_in_delta hd(opportunity.sell).quantity, 1899.0 / 13, 1.0e-7
    assert_in_delta opportunity.margin_percent, (1899.0 / 13 - 210) / 210 * 100, 0.001
    assert opportunity.profit.currency == @chaos
    assert_in_delta opportunity.profit.quantity, 1899.0 / 13 - 210, 1.0e-7
    # min(divine buy_leg_stock=80, chaos buy_leg_stock=50)
    assert opportunity.volume == 50
  end

  test "chaos leg's buy price collapses but its market sell price is still usable" do
    # Chaos-wisdom leg: price_at_lowest=1, price_at_highest=1.5 -> buy
    # floors to 0 (suggested_buy_price=nil), but market_sell_price=floor(1.5)=1
    # is still a valid, nonzero sell reference -- divine_to_chaos only
    # depends on the chaos leg's market_sell_price, not its (irrelevant,
    # collapsed) suggested_buy_price, so the opportunity must still compute.
    chaos_leg = snapshot(@chaos, @wisdom, {2, 2, 2, 3}, {40, 70, 1, 1})
    divine_leg = snapshot(@divine, @wisdom, {1, 50, 1, 80}, {30, 80, 1, 1})

    [opportunity] = BulkBuyOpportunityFinder.find([chaos_leg, divine_leg], @rate_210)

    assert hd(opportunity.start).currency == @divine
    assert hd(opportunity.sell).currency == @chaos
    # floor(80)-1
    assert hd(opportunity.via).quantity == 79.0
    # Sells through the chaos leg's market price (1, the floored raw
    # extreme with no +1 push), not the competitive 2 -- 79/1 = 79.
    assert_in_delta hd(opportunity.sell).quantity, 79.0, 1.0e-7
  end

  test "chaos leg price collapses below one: dropped safely, no crash" do
    # A real production incident: market_sell_price (no +1 push, unlike
    # suggested_sell_price) can floor to exactly 0 when its raw extreme is
    # below 1 -- e.g. both price_at_lowest=0.5 and price_at_highest=0.3
    # here. divine_to_chaos divides by chaos_leg.market_sell_price, so an
    # unguarded 0 would produce a crash rather than an empty result.
    chaos_leg = snapshot(@chaos, @wisdom, {2, 1, 10, 3}, {100, 50, 1, 1})
    divine_leg = snapshot(@divine, @wisdom, {1, 1700, 1, 1900}, {30, 80, 1, 1})

    assert BulkBuyOpportunityFinder.find([chaos_leg, divine_leg], @rate_210) == []
  end

  test "divine leg price collapses below one: dropped safely, no crash" do
    # divine_to_chaos uses divine_leg.suggested_buy_price as its via
    # amount -- when that collapses to nil (raw extreme below 1), the
    # opportunity must be dropped, not crash.
    chaos_leg = snapshot(@chaos, @deck, {1, 8, 1, 13}, {100, 50, 1, 1})
    divine_leg = snapshot(@divine, @deck, {2, 1, 10, 3}, {100, 50, 1, 1})

    assert BulkBuyOpportunityFinder.find([chaos_leg, divine_leg], @rate_210) == []
  end

  test "rate unavailable returns empty regardless of snapshots" do
    chaos_leg = snapshot(@chaos, @deck, {1, 8, 1, 13}, {100, 50, 1, 1})
    divine_leg = snapshot(@divine, @deck, {1, 1700, 1, 1900}, {30, 80, 1, 1})

    assert BulkBuyOpportunityFinder.find([chaos_leg, divine_leg], nil) == []
  end

  test "no matching divine leg for the same intermediary is not a candidate" do
    chaos_leg = snapshot(@chaos, @deck, {1, 8, 1, 13}, {100, 50, 1, 1})

    assert BulkBuyOpportunityFinder.find([chaos_leg], @rate_210) == []
  end

  test "only the chaos/divine reference pair itself excludes it as a candidate" do
    chaos_divine = snapshot(@chaos, @divine, {210, 1, 210, 1}, {50, 60, 50, 60})

    assert BulkBuyOpportunityFinder.find([chaos_divine], @rate_210) == []
  end

  test "non-finite ratio on one leg drops that intermediary entirely" do
    # lowest_ratio_a=0
    chaos_leg = snapshot(@chaos, @deck, {0, 8, 1, 13}, {100, 50, 1, 1})
    divine_leg = snapshot(@divine, @deck, {1, 1700, 1, 1900}, {30, 80, 1, 1})

    assert BulkBuyOpportunityFinder.find([chaos_leg, divine_leg], @rate_210) == []
  end
end
